### PR TITLE
feat: add transaction contextmanager (CAP-725)

### PR DESCRIPTION
**Pull request by floredelasteyrie**

### CODE REVIEW

File: src/outcome/utils/transactions.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    ops.add(create_user)
+    op.apply()
+```
+
+"""
+
+from contextlib import contextmanager
+from enum import Enum
+
+from transaction import Transaction, TransactionManager
```
Comment (floredelasteyrie) [inline src/outcome/utils/transactions.py:26]: @d-tw do we have to do something special as the import will be optional?

File: make/setup.Makefile [hunk 2]
```diff
@@ -26,15 +26,15 @@ install-build-system: ## Install poetry
...
 
 dev-setup: install-build-system ## Install the dependencies for dev environments
 	./pre-commit.sh
-	LDFLAGS="-L$$(brew --prefix openssl)/lib" poetry install
+	LDFLAGS="-L$$(brew --prefix openssl)/lib" poetry install --extras all
 
 production-setup: install-build-system ## Install the dependencies for production environments
 	poetry config virtualenvs.create false
-	poetry install --no-dev --no-interaction --no-ansi
+	poetry install --no-dev --no-interaction --no-ansi --extras all
```
Comment (d-tw) [inline make/setup.Makefile:38]: It's better to put the mark the dependencies as development dependencies

File: pyproject.toml [hunk 3]
```diff
@@ -26,6 +26,10 @@ requests = "^2.24.0"
 cachetools = "^4.1.1"
 pyfakefs = "^4.1.0"
 python-dotenv = "^0.15.0"
+transaction = { version="^3.0.0", optional = true }
+
+[tool.poetry.extras]
+all = ["transaction"]
```
Comment (d-tw) [inline pyproject.toml:32]:
```suggestion
transaction = ["transaction"]
```

File: src/outcome/utils/transactions.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    ops.add(op)
+```
+
+- We can disable auto-apply if necessary:
+```
+op = None
+with transaction(auto_apply=False) as ops:
+    op = Operation(apply_fn=your_apply_fn, rollback_fn=your_rollback_fn)
+    ops.add(create_user)
+    op.apply()
```
Comment (d-tw) [inline src/outcome/utils/transactions.py:18]: Maybe when auto apply is set to false, the context manager should check if all of the operations were applied, and raise an exception if some operations were skipped

File: src/outcome/utils/transactions.py (hunk 1 above)
Comment (d-tw) [reply src/outcome/utils/transactions.py:26]: You can always make it a bit more dev-friendly:

```suggestion
try:
    from transaction import Transaction, TransactionManager
except ImportError as ex:
    raise Exception('You need to install the 'transaction' extra to use the transaction features') 
```

File: src/outcome/utils/transactions.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    errored = 'errored'
+    reset = 'reset'
+
+
+class InvalidStateForApply(Exception):
+    ...
+
+
+class InvalidStateForRollback(Exception):
+    ...
```
Comment (d-tw) [inline src/outcome/utils/transactions.py:56]: This can just be one Exception

**Review by d-tw — CHANGES_REQUESTED** [review]

File: src/outcome/utils/transactions.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    def apply(self):
+        for op in self.operations:
+            op.apply()
+
+    def abort(self, txn):
+        for op in reversed(self.operations):
+            if op.state == OperationState.applied:
+                op.rollback()
+
+    def check_applied(self):
```
Comment (d-tw) [inline src/outcome/utils/transactions.py:110]:
```suggestion
    def ensure_applied(self):
```

File: src/outcome/utils/transactions.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        txn.join(self)
+
+    def add(self, operation: Operation):
+        self.operations.append(operation)
+
+    def apply(self):
+        for op in self.operations:
+            op.apply()
+
+    def abort(self, txn):
```
Comment (d-tw) [inline src/outcome/utils/transactions.py:105]:
```suggestion
    def abort(self, txn: Transaction):
```

File: src/outcome/utils/transactions.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+try:
+    from transaction import Transaction, TransactionManager  # noqa: WPS433 - nested import
+    from transaction.interfaces import IDataManager  # noqa: WPS433 - nested import
+except ImportError as exc:
+    raise Exception('You need to install the `transaction` extra to use the transaction features') from exc
+
+
+@contextmanager
+def transaction(auto_apply=True):
```
Comment (d-tw) [inline src/outcome/utils/transactions.py:37]:
```suggestion
def transaction(auto_apply: bool =True):
```

**Review by d-tw — CHANGES_REQUESTED** [review]

**Review by d-tw — APPROVED** [review]

**Pull request merged**